### ISSUE TITLE
Cannot compile project using cross to OpenBSD i386 on Linux x64

### ISSUE THREAD

**Issue by Alexey-T** [bgrabitmap/bgrabitmap#216]
```
Verbose: Assembling bgraanimatedgif
Error: /home/user/cuda/bgrabitmap/bgrabitmap/lib/i386-openbsd-gtk2-3.2.3/bgracolorint.s: Assembler messages:
Error: /home/user/cuda/bgrabitmap/bgrabitmap/lib/i386-openbsd-gtk2-3.2.3/bgracolorint.s:442: Error: no instruction mnemonic suffix given and no register operands; can't size instruction
Error: /home/user/cuda/bgrabitmap/bgrabitmap/lib/i386-openbsd-gtk2-3.2.3/bgracolorint.s:460: Error: no instruction mnemonic suffix given and no register operands; can't size instruction
Error: /home/user/cuda/bgrabitmap/bgrabitmap/lib/i386-openbsd-gtk2-3.2.3/bgracolorint.s:478: Error: no instruction mnemonic suffix given and no register operands; can't size instruction
Error: /home/user/cuda/bgrabitmap/bgrabitmap/lib/i386-openbsd-gtk2-3.2.3/bgracolorint.s:496: Error: no instruction mnemonic suffix given and no register operands; can't size instruction
Error: /home/user/cuda/bgrabitmap/bgrabitmap/lib/i386-openbsd-gtk2-3.2.3/bgracolorint.s:699: Error: no instruction mnemonic suffix given and no register operands; can't size instruction
Error: /home/user/cuda/bgrabitmap/bgrabitmap/lib/i386-openbsd-gtk2-3.2.3/bgracolorint.s:732: Error: no instruction mnemonic suffix given and no register operands; can't size instruction
Error: /home/user/cuda/bgrabitmap/bgrabitmap/lib/i386-openbsd-gtk2-3.2.3/bgracolorint.s:765: Error: no instruction mnemonic suffix given and no register operands; can't size instruction
Error: /home/user/cuda/bgrabitmap/bgrabitmap/lib/i386-openbsd-gtk2-3.2.3/bgracolorint.s:798: Error: no instruction mnemonic suffix given and no register operands; can't size instruction
Verbose: Compiling bgracolorint.pas
Verbose: Assembling bgracolorint
bgracolorint.pas(364,0) Error: Error while assembling exitcode 1
bgracolorint.pas(364,0) Verbose: There were 2 errors compiling module, stopping
Verbose: Compilation aborted
Verbose: /home/user/fpcupdeluxe/fpcupdeluxe32/fpc/bin/x86_64-linux/ppcross386 returned an error exitcode
```

FPC 3.2-fixes. Lazarus 3.99.

**Comment by fredvs** [comment]
Hello.

May I ask you if you are able to compile-run Lazarus applications on OpenBSD?
Are you able to compile+run CudaText for OpenBSD?

With msegui, the apps can compile but for running, it must be done via terminal, 2Xclick on the binary crash the application.

But this is for OpenBSD x86_64, I never try OpenBSD i386 on msegui.
Also the compilation was done on a VM machine, not cross-compiling.

**Comment by Alexey-T** [comment]
before I update Deluxe to 2.4.0 I had Deluxe 2.2.0s, and cross to OpenBSD x64 worked! From the same PC with Linux x64.

Then author added cross to OpenBSD i386 and it fails.

**Comment by fredvs** [comment]
Ok, for compiling but for running, did you try to run it on a VM (or true) OpenBSD machine?

Does it run when 2xclick on the binary?

For OpenBSD i386, I have to install it on a VM machine and see what it says.

Out of the subject, I love their idea to give, for each new release, a song!
https://www.openbsd.org/lyrics.html

**Comment by fredvs** [comment]
@Alexey-T  did you try to compile+run CudaText with NetBSD?
msegui apps compile now but there was lot of change to make image/bitmap things to work. (at the moment gray-filter still does not work ok here).

**Comment by Alexey-T** [comment]
>Does it run when 2xclick on the binary?

yes. Ubuntu 20.4 with VMWare Workstation 15. VM with OpenBSD 7.2, double-click on Thunar file manager runs CudaText 1.186.0.

**Comment by Alexey-T** [comment]
> did you try to compile+run CudaText with NetBSD?

yes, same method as OpenBSD, on VMware Workstation 15 VM.

**Comment by fredvs** [comment]
>Yes. Ubuntu 20.4 with VMWare Workstation 15. VM with OpenBSD 7.2, double-click on Thunar file manager on 'cudatext' runs CudaText 1.186.0.

Ha, ok perfect.

So some more white nights needed to fix this.
(And I dont know where to ask, Lazarus forum did not answer, there are no OpenBSD git-issues and I did not find a OpenBSD forum.  It must be something missing for the ELF but I dont catch what.  When running from 2xclick, the application freeze and the CPU is working 100% without decrease, the application must be killed.
And strangely, running the app from terminal all is ok.
If somebody has a idea what could be the problem or what to ask in internet, it is welcome.

For NETBsd, the same (with still some problems with image+bitmap).

Anyway many thanks for the infos.

And for OpenBSD i386 + BGRABitmap sorry I cannot test now, I will install it asap.

**Comment by fredvs** [comment]
@Alexey-T : Maybe not concerned but in the error messages you get, it seems to come from assembler call.

Can you try to compile commenting in unit bgracolorint.pas, line 6:

```
{$mode objfpc}{$H+}
{$ifdef CPUI386}
  // {$define BGRACOLORINT_USEASM} // comment this line
{$endif}
{$ifdef DARWIN}
  {$undef BGRACOLORINT_USEASM}
{$endif}
```

**Comment by Alexey-T** [comment]
this is bad fix, because it slows down the component on Windows i386.
i guess component will work, though.

**Comment by fredvs** [comment]
>this is bad fix, because it slows down the component on Windows i386.

Of course, it is only to test.

If you prefer, you can use this:

```
{$if defined(i386) and not defined(openbsd)}
  {$define BGRACOLORINT_USEASM} 
{$endif}
{$ifdef DARWIN}
  {$undef BGRACOLORINT_USEASM}
{$endif}
```

**Comment by Alexey-T** [comment]
```
{$if defined(i386) and not defined(openbsd)}
  {$define BGRACOLORINT_USEASM} 
{$endif}
```
with this, component is compiled.

**Comment by fredvs** [comment]
It seems that OpenBSD has his own ASM clang compiler with own custom methods...
By the way, does it make a big difference to use ASM code for color-assign in Unix OS?

**Comment by fredvs** [comment]
Hello.

About the problem with msegui on NetBSD and OpenBSD, when 2xclick on the binary freezed the app, it is fixed now in last commit: https://github.com/mse-org/mseide-msegui/commit/2e1f3b55c048d41ea193a40b604e05efc6365b12

So sorry for the noise (and thanks Alexey to have try with CudaText).

**Comment by circular17** [comment]
Hello,

So if I understand, you would advise to add `and not defined(openbsd)` to the directive?

**Comment by fredvs** [comment]
Yes or you have to find the equivalent ASM code for OpenBSD.

**Comment by circular17** [comment]
Ok.

Isn't it ```{$if defined(CPUI386) and not defined(openbsd)}``` rather than ```{$if defined(i386) and not defined(openbsd)}```?

**Comment by fredvs** [comment]
Ha, yes, sometimes defined(i386) works, sometime no.

For example, in fpc/compiler/x86/cpubase.pas there is:

```
{$if defined(x86_64)}
        {$i r8664stab.inc}
{$elseif defined(i386)}
        {$i r386stab.inc}
{$elseif defined(i8086)}
        {$i r8086stab.inc}
{$endif}
      );
```

But yes, better to use defined(**CPU**I386)
